### PR TITLE
implement tx_lookup stage

### PR DESCRIPTION
**Pull request by blasrodri**
Got some pointers from @Giulio2002.
Please guide me through on what's wrong/missing.
Thanks!

### CODE REVIEW

File: src/stages/tx_lookup.rs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        let start_key = past_progress.to_be_bytes();
+        let mut collector = Collector::new(OPTIMAL_BUFFER_CAPACITY);
+        let walker = txdb::walk(&mut bodies_cursor, &start_key, 0);
+        pin!(walker);
+
+        while let Some((block_key, ref body)) = walker.try_next().await? {
+            let (block_number, block_hash) = (&block_key[0..8], &block_key[8..]);
+
+            let body = rlp::decode::<BodyForStorage>(body)?;
+            for ref uncle in body.uncles {
```
Comment (Giulio2002) [inline src/stages/tx_lookup.rs:51]: Why are you taking uncles, you need to take transactions

File: src/stages/tx_lookup.rs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+                let bytes = rlp::encode(uncle);
+                let hashed_tx_data = hash_data(bytes.as_ref());
+                collector.collect(Entry {
+                    key: hashed_tx_data.as_bytes().to_vec(),
+                    value: [block_number, block_hash].concat(),
```
Comment (Giulio2002) [inline src/stages/tx_lookup.rs:56]: value is just big endian block number, no need for concatenation

File: src/stages/tx_lookup.rs (hunk 1 above)
Comment (blasrodri) [reply src/stages/tx_lookup.rs:51]: Where are the txs?

Comment (Giulio2002) [reply src/stages/tx_lookup.rs:51]: you need to use the transaction base id and transaction count from the block body you computed to extract them from the EthTx bucket/table/cursor

Comment (Giulio2002) [reply src/stages/tx_lookup.rs:51]: I would suggest taking a read on the ReadTransactions function in Erigon

**Review by Giulio2002 — CHANGES_REQUESTED** [review]

File: src/stages/tx_lookup.rs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        (U64::from_big_endian(last_processed_block_number.as_ref()) + 1)
+            .to_big_endian(&mut start_block_number);
+
+        let walker_block_body = txdb::walk(&mut bodies_cursor, &start_block_number, 0);
+        pin!(walker_block_body);
+
+        while let Some((block_body_key, ref block_body_value)) =
+            walker_block_body.try_next().await?
+        {
+            let block_number = &block_body_key[..8];
```
[the remark this answers is not in the document]
Comment (Giulio2002) [reply src/stages/tx_lookup.rs:67]: this behaviour is just for this stage, you wont need it for the others

**Comment by Giulio2002** [comment]
LGTM @vorot93 , could you look at this before merging

**Review by Giulio2002 — APPROVED** [review]

**Comment by Giulio2002** [comment]
in the meanwhile could you had some tests for this stage?

File: src/stages/tx_lookup.rs [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+            while let Some((_tx_key, ref tx_value)) = walker_block_txs.try_next().await? {
+                if num_txs > tx_count {
+                    break;
+                }
+
+                let hashed_tx_data = hash_data(tx_value);
+                collector.collect(Entry {
+                    key: hashed_tx_data.as_bytes().to_vec(),
+                    value: block_number.clone(),
+                    id: 0, // ?
```
Comment (vorot93) [inline src/stages/tx_lookup.rs:90]: @Giulio2002

Comment (Giulio2002) [reply src/stages/tx_lookup.rs:90]: the id does not matter, just put 0, the is only used when we need to use the data provider

**Comment by vorot93** [comment]
Thanks!

**Pull request merged**